File: trunk/avr/vm/interpreter.c
#include "interpreter.h"
#include "vmstatus.h"

#include <stdint.h>
#include <string.h>

enum {
	RET,
	RET_N,
	RETI,
	RETI_N,
	LIT,
	SETREG,
	GETREG,
	SETTIMER_S,
	SETTIMER_MS,
	STOPTIMER,
	QBRANCH,
	BRANCH,
	ADD,
	SUB,
	MUL,
	DIV,
	NEGATE,
	OR,
	AND,
	NOT,
	LT,
	LE,
	GT,
	GE,
	EQ,
	NEQ,
	DROP,
	INC,
	DEC,
	CALL,
	RESERVED,
	MULDIV,
	GETLOCAL,
	SETLOCAL,
	ENTER,
	PRINT,
	SETEXTREG,
	GETEXTREG,
	THROW,
};

#define GETREGSHORT 0x80
#define SETREGSHORT 0xC0
#define LITSHORT 0x40

static uint16_t codeOffset;
static uint16_t events;
static uint16_t timers;
static uint16_t funcs;
static uint16_t TCBs;
static int ntimers;
static int nevents;
static int nfuncs;
static int nTCB;

#define MAX_TIMERS 16
#define MAX_EVENTS 16

static int32_t vmTimerCnts[MAX_TIMERS];
static int16_t vmEventStates[MAX_EVENTS];

static uint16_t fp;
static uint16_t ip;

static int16_t vmExec(uint16_t ip);

static uint16_t getUint16(uint16_t addr) {
	return (vmReadByte(addr) << 8) | vmReadByte(addr + 1);
}

static uint16_t vmGetTimerHandler(uint8_t timer) {
	return getUint16(timers + timer * 2);
}

static uint16_t vmGetEventCondition(uint8_t event) {
	return getUint16(events + event * 4);
}

static uint16_t vmGetEventHandler(uint8_t event) {
	return getUint16(events + event * 4 + 2);
}

static uint16_t vmGetFuncCode(uint8_t func) {
	return getUint16(funcs + func * 2);
}

int getUIStart() {
	return 0;
}

int getUIEnd() {
	uint16_t ptr = 0;
	int nUI = vmReadByte(ptr);
	ptr++;
	while (nUI--) {
		ptr++; // reg
		ptr++; // type
		ptr += vmReadByte(ptr) + 1; // text
	}
	return ptr;
}

static void initVars() {
	uint16_t ptr = getUIEnd();
	nevents = vmReadByte(ptr);
	ptr++;
	events = ptr;
	ptr += nevents * 4;

	ntimers = vmReadByte(ptr);
	ptr++;
	timers = ptr;
	ptr += ntimers * 2;

	nfuncs = vmReadByte(ptr);
	ptr++;
	funcs = ptr;
	ptr += nfuncs * 2;

	nTCB = vmReadByte(ptr);
	ptr++;
	TCBs = ptr;
	ptr += nTCB * 6;

	codeOffset = ptr;

	stackPtr = stack + STACK_SIZE;

	memset(vmTimerCnts, 0, sizeof(vmTimerCnts));
	memset(vmEventStates, 0, sizeof(vmEventStates));
}

void vmInit() {
	if (vmGetStatus() != VMSTATUS_STOPPED)
		return;
	initVars();
	vmSetStatus(VMSTATUS_RUNNING);
	vmExec(vmGetFuncCode(0) + codeOffset);
	vmSetStatus(VMSTATUS_STOPPED);
}

static void vmMain() {
	initVars();
	vmSetStatus(VMSTATUS_RUNNING);
	vmExec(vmGetFuncCode(1) + codeOffset);
}

static int16_t eval(uint16_t ip) {
	if (vmExec(ip) == 0)
		return vmPop();
	return 0;
}

void vmStep(int ms) {
	int i;

	if (vmGetStatus() != VMSTATUS_RUNNING)
		return;
	for (i = 0; i < nevents; i++) {
		uint16_t cond = vmGetEventCondition(i);
		int change = (cond & 0x8000) != 0;
		cond &= 0x7FFF;
		int16_t newState = eval(cond + codeOffset);
		if (vmGetStatus() != VMSTATUS_RUNNING)
			return;
		if ((!change && !vmEventStates[i] && newState)
				|| (change && vmEventStates[i] != newState)) {
			vmExec(vmGetEventHandler(i) + codeOffset);
			if (vmGetStatus() != VMSTATUS_RUNNING)
				return;
		}
		vmEventStates[i] = newState;
	}

	for (i = 0; i < ntimers; i++) {
		if (vmTimerCnts[i]) {
			vmTimerCnts[i] -= ms;
			if (vmTimerCnts[i] <= 0) {
				vmTimerCnts[i] = 0;
				vmExec(vmGetTimerHandler(i) + codeOffset);
				if (vmGetStatus() != VMSTATUS_RUNNING)
					return;
			}
		}
	}
}

static uint16_t findCatchBlock(uint16_t ip) {
	uint16_t bestHandler = 0;
	uint16_t bestFrom = 0;
	int i;

	for (i = 0; i < nTCB; i++) {
		uint16_t tcb = TCBs + i * 6;
		uint16_t from = getUint16(tcb);
		uint16_t to = getUint16(tcb + 2);
		if (ip >= from && ip < to && from > bestFrom) {
			bestFrom = from;
			bestHandler = getUint16(tcb + 4);
		}
	}

	return bestHandler;
}

static void call(uint16_t addr) {
	vmPush(ip);
	vmPush(fp);
	fp = stackPtr - stack;
	ip = addr;
}

static void ret() {
	vmSetStack(stack + fp);
	fp = vmPop();
	ip = vmPop();
}

static void throwException(int16_t e) {
	for (;;) {
		vmSetStack(stack + fp);
		uint16_t catchBlock = findCatchBlock(ip - codeOffset);
		if (catchBlock != 0) {
			vmPush(e);
			ip = catchBlock + codeOffset;
			break;
		} else {
			ret();
			if (ip == 0)
				return;
		}
	}
}

#define LOCAL(n) (stack[fp + 2 + (n)])

static int16_t vmExec(uint16_t _ip) {

	stackPtr = stack + STACK_SIZE;
	ip = 0;
	fp = 0;

	call(_ip);

	for (;;) {
		if (vmGetStatus() != VMSTATUS_RUNNING)
			return -1;

		uint8_t c = vmReadByte(ip++);

		if ((c & 0xC0) == (GETREGSHORT & 0xFF)) {
			vmPush(_getReg(c & 0x3F));
			continue;
		} else if ((c & 0xC0) == (SETREGSHORT & 0xFF)) {
			_setReg(c & 0x3F, vmPop());
			continue;
		} else if ((c & 0xC0) == (LITSHORT & 0xFF)) {
			int16_t v = (int16_t) (((int8_t) ((c & 0x3F) << 2)) >> 2);
			vmPush(v);
			continue;
		}

		switch (c) {
		case CALL: {
			uint16_t addr = vmGetFuncCode(vmReadByte(ip++)) + codeOffset;
			call(addr);
			break;
		}
		case RET:
			ret();
			if (ip == 0)
				return 0;
			break;
		case RET_N: {
			uint8_t n = vmReadByte(ip);
			ret();
			vmChangeStack(-n);
			if (ip == 0)
				return 0;
			break;
		}
		case RETI: {
			int16_t res = vmPop();
			ret();
			vmPush(res);
			if (ip == 0)
				return 0;
			break;
		}
		case RETI_N: {
			uint8_t n = vmReadByte(ip);
			int16_t res = vmPop();
			ret();
			vmChangeStack(-n);
			vmPush(res);
			if (ip == 0)
				return 0;
			break;
		}
		case THROW: {
			int16_t res = vmPop();
			throwException(res);
			if (ip == 0)
				return res;
			break;
		}
		case GETLOCAL: {
			int n = (int8_t) vmReadByte(ip++);
			vmPush(LOCAL(n));
			break;
		}
		case SETLOCAL: {
			int n = (int8_t) vmReadByte(ip++);
			LOCAL(n) = vmPop();
			break;
		}
		case ENTER: {
			uint16_t link = vmPop();
			uint16_t ret = vmPop();
			int n = vmReadByte(ip++);
			vmChangeStack(n);
			vmPush(ret);
			vmPush(link);
			fp = stackPtr - stack;
			break;
		}
		case LIT:
			vmPush(getUint16(ip));
			ip += 2;
			break;

		case SETREG:
			_setReg(vmReadByte(ip++), vmPop());
			break;
		case GETREG:
			vmPush(_getReg(vmReadByte(ip++)));
			break;
		case SETTIMER_S:
			vmTimerCnts[vmReadByte(ip++)] = (int32_t) vmPop() * 1000;
			break;
		case SETTIMER_MS:
			vmTimerCnts[vmReadByte(ip++)] = vmPop();
			break;
		case STOPTIMER:
			vmTimerCnts[vmReadByte(ip++)] = 0;
			break;
		case QBRANCH:
			if (vmPop())
				ip++;
			else
				ip += (int8_t) vmReadByte(ip) + 1;
			break;
		case BRANCH:
			ip += (int8_t) vmReadByte(ip) + 1;
			break;
		case ADD:
			vmPush(vmPop() + vmPop());
			break;
		case SUB: {
			int16_t r = vmPop();
			vmPush(vmPop() - r);
			break;
		}
		case MUL:
			vmPush(vmPop() * vmPop());
			break;
		case DIV: {
			int16_t r = vmPop();
			if (r == 0) {
				throwException(ARITHMETIC_EXCEPTION);
				if (ip == 0)
					return ARITHMETIC_EXCEPTION;
			} else {
				vmPush(vmPop() / r);
			}
			break;
		}
		case OR: {
			int n1 = vmPop();
			int n2 = vmPop();
			vmPush(n1 || n2);
			break;
		}
		case AND: {
			int n1 = vmPop();
			int n2 = vmPop();
			vmPush(n1 && n2);
			break;
		}
		case NOT:
			vmPush(!vmPop());
			break;
		case NEGATE: {
			vmPush(-vmPop());
			break;
		}
		case LT: {
			int16_t r = vmPop();
			vmPush(vmPop() < r);
			break;
		}
		case LE: {
			int16_t r = vmPop();
			vmPush(vmPop() <= r);
			break;
		}
		case GT: {
			int16_t r = vmPop();
			vmPush(vmPop() > r);
			break;
		}
		case GE: {
			int16_t r = vmPop();
			vmPush(vmPop() >= r);
			break;
		}
		case EQ:
			vmPush(vmPop() == vmPop());
			break;
		case NEQ:
			vmPush(vmPop() != vmPop());
			break;
		case DROP:
			vmPop();
			break;
		case PRINT:
			vmPrintInt(vmPop());
			break;
		case INC: {
			uint8_t reg = vmReadByte(ip++);
			_setReg(reg, _getReg(reg) + 1);
			break;
		}
		case DEC: {
			uint8_t reg = vmReadByte(ip++);
			_setReg(reg, _getReg(reg) - 1);
			break;
		}
		case MULDIV: {
			int32_t n3 = vmPop();
			int32_t n2 = vmPop();
			int32_t n1 = vmPop();
			if (n3 == 0) {
				throwException(ARITHMETIC_EXCEPTION);
				if (ip == 0)
					return ARITHMETIC_EXCEPTION;
			} else {
				vmPush(n1 * n2 / n3);
			}
			break;
		}
		default:
			vmSetStatus(VMSTATUS_INVALID_BYTECODE);
			break;
		}
	}

	return -1;
}

void vmStart(int8_t b) {
	if (b) {
		vmMain();
	} else {
		vmSetStatus(VMSTATUS_STOPPED);
	}
}

File: trunk/zavorovo/vm/src/kvv/evlang/rt/Interpreter.java
package kvv.evlang.rt;

import java.util.List;

public abstract class Interpreter {

	private final RTContext context;

	public Interpreter(RTContext cont) {
		this.context = cont;
	}

	public abstract void setExtReg(int addr, int reg, int value);

	public abstract int getExtReg(int addr, int reg);

	static private class Stack {
		private int[] data = new int[100];
		private int sp = data.length;

		public boolean isEmpty() {
			return sp == data.length;
		}

		public int pop() {
			return data[sp++];
		}

		public void push(int n) {
			data[--sp] = n;
		}

		public int getSP() {
			return sp;
		}

		public int getAt(int off) {
			return data[off];
		}

		public void setAt(int off, int n) {
			data[off] = n;
		}
	}

	private Stack stack = new Stack();
	private int ip;
	private int fp;

	public void interpret(int off) throws UncaughtExceptionException {
		if (!stack.isEmpty())
			throw new IllegalStateException();
		if (ip != 0)
			throw new IllegalStateException();
		if (fp != 0)
			throw new IllegalStateException();
		_interpret(off);
		if (!stack.isEmpty())
			throw new IllegalStateException();
	}

	public int eval(int off) throws UncaughtExceptionException {
		if (!stack.isEmpty())
			throw new IllegalStateException();
		if (ip != 0)
			throw new IllegalStateException();
		if (fp != 0)
			throw new IllegalStateException();
		_interpret(off);
		int res = stack.pop();
		if (!stack.isEmpty())
			throw new IllegalStateException();
		return res;
	}

	private void call(int addr) {
		stack.push(ip);
		stack.push(fp);
		fp = stack.getSP();
		ip = addr;
	}

	private void ret() {
		stack.sp = fp;
		fp = stack.pop();
		ip = stack.pop();
	}

	private void throwException(int e) throws UncaughtExceptionException {
		for (;;) {
			stack.sp = fp;
			TryCatchBlock tcb = context.findTryCatchBlock(ip);
			if (tcb != null) {
				stack.push(e);
				ip = tcb.handler;
				break;
			} else {
				ret();
				if (ip == 0)
					throw new UncaughtExceptionException();
			}
		}
	}

	private void _interpret(int _ip)
			throws UncaughtExceptionException {
		List<Byte> code = context.codeArr;

		call(_ip);

		while (true) {
			int right;
			int left;
			int off;
			int timer;

			byte c = code.get(ip++);
			if ((c & 0xC0) == BC.GETREGSHORT) {
				int reg = c & 0x3F;
				stack.push(context.regs[reg]);
				continue;
			} else if ((c & 0xC0) == BC.SETREGSHORT) {
				int reg = c & 0x3F;
				context.regs[reg] = stack.pop();
				continue;
			} else if ((c & 0xC0) == BC.LITSHORT) {
				int val = (c & 0x3F) << 26 >> 26;
				stack.push(val);
				continue;
			}

			BC bc = BC.values()[c];
			switch (bc) {
			case CALL:
				int addr = context.funcs[code.get(ip++)].code;
				call(addr);
				break;
			case RET:
				ret();
				if (ip == 0)
					return;
				break;
			case RET_N:
				int n = code.get(ip++);
				ret();
				while (n-- > 0)
					stack.pop();
				if (ip == 0)
					return;
				break;
			case RETI:
				int res = stack.pop();
				ret();
				stack.push(res);
				if (ip == 0)
					return;
				break;
			case RETI_N:
				n = code.get(ip++);
				res = stack.pop();
				ret();
				while (n-- > 0)
					stack.pop();
				stack.push(res);
				if (ip == 0)
					return;
				break;

			case THROW:
				res = stack.pop();
				throwException(res);
				break;

			case GETLOCAL:
				n = code.get(ip++);
				stack.push(stack.getAt(fp + 2 + n));
				break;
			case SETLOCAL:
				n = code.get(ip++);
				stack.setAt(fp + 2 + n, stack.pop());
				break;
			case ENTER:
				int link = stack.pop();
				int ret = stack.pop();
				n = code.get(ip++);
				while (n-- > 0)
					stack.push(0);
				stack.push(ret);
				stack.push(link);
				fp = stack.sp;
				break;
			case ADD:
				stack.push(stack.pop() + stack.pop());
				break;
			case SUB:
				stack.push(-(stack.pop() - stack.pop()));
				break;
			case MUL:
				stack.push(stack.pop() * stack.pop());
				break;
			case DIV:
				right = stack.pop();
				left = stack.pop();
				if (right == 0)
					throwException(Exc.ARITHMETIC_EXCEPTION.ordinal());
				else
					stack.push(left / right);
				break;
			case AND: {
				int n1 = stack.pop();
				int n2 = stack.pop();
				stack.push((n1 != 0 && n2 != 0) ? 1 : 0);
				break;
			}
			case OR: {
				int n1 = stack.pop();
				int n2 = stack.pop();
				stack.push((n1 != 0 || n2 != 0) ? 1 : 0);
				break;
			}
			case EQ:
				stack.push((stack.pop() == stack.pop()) ? 1 : 0);
				break;
			case NEQ:
				stack.push((stack.pop() != stack.pop()) ? 1 : 0);
				break;
			case GT:
				right = stack.pop();
				stack.push((stack.pop() > right) ? 1 : 0);
				break;
			case GE:
				right = stack.pop();
				stack.push((stack.pop() >= right) ? 1 : 0);
				break;
			case LT:
				right = stack.pop();
				stack.push((stack.pop() < right) ? 1 : 0);
				break;
			case LE:
				right = stack.pop();
				stack.push((stack.pop() <= right) ? 1 : 0);
				break;
			case LIT:
				byte hi = code.get(ip++);
				byte lo = code.get(ip++);
				stack.push((hi << 8) + (lo & 0xFF));
				break;
			case NOT:
				stack.push(stack.pop() != 0 ? 0 : 1);
				break;
			case NEGATE:
				stack.push(-stack.pop());
				break;
			case GETREG:
				int reg = code.get(ip++) & 0xFF;
				stack.push(context.regs[reg]);
				break;
			case SETREG:
				int reg1 = code.get(ip++) & 0xFF;
				context.regs[reg1] = stack.pop();
				break;
			case GETEXTREG:
				addr = code.get(ip++) & 0xFF;
				reg = code.get(ip++) & 0xFF;
				stack.push(getExtReg(addr, reg));
				break;
			case SETEXTREG:
				addr = code.get(ip++) & 0xFF;
				reg = code.get(ip++) & 0xFF;
				setExtReg(addr, reg, stack.pop());
				break;
			case BRANCH:
				off = code.get(ip++);
				ip += off;
				break;
			case QBRANCH:
				off = code.get(ip++);
				if (stack.pop() == 0)
					ip += off;
				break;
			case DROP:
				stack.pop();
				break;
			case PRINT:
				System.out.print(stack.pop() + " ");
				break;
			case SETTIMER_MS:
				timer = code.get(ip++);
				context.timers[timer].cnt = stack.pop();
				break;
			case SETTIMER_S:
				timer = code.get(ip++);
				context.timers[timer].cnt = stack.pop() * 1000;
				break;
			case STOPTIMER:
				timer = code.get(ip++);
				context.timers[timer].cnt = 0;
				break;
			case INC:
				reg = code.get(ip++) & 0xFF;
				context.regs[reg]++;
				break;
			case DEC:
				reg = code.get(ip++) & 0xFF;
				context.regs[reg]--;
				break;
			case MULDIV: {
				int n3 = stack.pop();
				int n2 = stack.pop();
				int n1 = stack.pop();
				if (n3 == 0)
					throwException(Exc.ARITHMETIC_EXCEPTION.ordinal());
				else
					stack.push(n1 * n2 / n3);
				break;
			}
			default:
				throw new RuntimeException("unknown bytecode " + c);
			}
		}

	}
}
